lib/masdb/data_map.ex

```diff
@@ -0,0 +1,140 @@
+defmodule Masdb.Data.Val do
+  @type time :: Masdb.Timestamp.t
+  @type value :: integer | float | boolean | String.t | :unknown
+  
+  @type t :: %Masdb.Data.Val{
+    since_ts: time,
+    value: value,
+  }
+  @enforce_keys [:since_ts, :value]
+  defstruct [:since_ts, :value]
+end
+
+defmodule Masdb.Data.Row do
+  @type col_name :: String.t
+  @type columns :: %{optional(col_name) => list(Masdb.Data.Val.t)}
+
+  @type t :: %Masdb.Data.Row{columns: columns}
+  @enforce_keys []
+  defstruct [columns: %{}]
+end
+
+defmodule Masdb.Data.Table do
+  @type row_id :: String.t
+  @type rows :: %{optional(row_id) => Masdb.Data.Row.t}
+
+  @type t :: %Masdb.Data.Table{rows: rows}
+  @enforce_keys []
+  defstruct [rows: %{}]
+end
+
+defmodule Masdb.Data.Map do
+  use Pipe
+
+  @type schema_name :: String.t
+  @type node_id :: String.t
+  @type time :: Masdb.Timestamp.t
+  @type opened_time :: time | nil
+  @type data_map :: %{optional(schema_name) => Masdb.Data.Table.t}
+
+  @type t :: %Masdb.Data.Map{
+    node_id: node_id,
+    last_update_time: time,
+    last_sync_time: opened_time,
+    map: data_map
+  }
+  @enforce_keys [:node_id, :last_update_time]
+  defstruct [:node_id, :last_update_time, last_sync_time: nil, map: %{}]
+
+  def insert(%Masdb.Data.Map{ node_id: n_id, last_sync_time: l_s_t, map: old_map }, schema, values) do
+    timestamp = Masdb.Timestamp.get_timestamp()
+    row_id = n_id <> to_string(timestamp.unique_integer)
+    old_rows = Map.get(old_map, schema.name, %Masdb.Data.Table{}).rows
+    pks = Masdb.Schema.get_pk(schema)
+    non_nullables = Masdb.Schema.get_non_nullables(schema)
+
+    {row_id, pipe_matching({:ok, _},
+              {:ok, values}
+              |> validate_with_schema(schema)
+              |> validate_nullable_referenced(non_nullables)
+              |> validate_pk_uniqueness(old_rows, pks)
+              |> normalize_to_row(timestamp)
+              |> put_new_row(schema.name, row_id, n_id, l_s_t, old_map, timestamp))}
+  end
+
+  defp validate_with_schema({:ok, values}, schema) do
+    keys = Map.keys(values)
+    case length(keys) do
+      0 -> {:cannot_insert_empty_row, values}
+      _ -> {validate_col_exists(keys, schema), values}
+    end
+  end
+
+  defp validate_nullable_referenced(values, []) do
+    values
+  end
+
+  defp validate_nullable_referenced({:ok, values}, [nn_nullable | other_nn_nullables]) do
+    case Map.get(values, nn_nullable) do
+      nil -> {:non_nullable_not_referenced, values}
+        _ -> validate_nullable_referenced({:ok, values}, other_nn_nullables)
+    end
+  end
+
+  defp validate_pk_uniqueness({:ok, values}, %Masdb.Data.Table{}, _) do
```

Comment:
You are right, this one wasn't usefull, I wanned to simplify the call when the table is empty, but it's useless